lib/Onyx.d.ts

```diff
@@ -0,0 +1,243 @@
+ * @returns {Promise}
+ */
+declare function multiSet<K extends KeyList>(data: Record<K, Values[K]>): Promise<void>;
+
+/**
+ * Merge a new value into an existing value at a key.
+ *
+ * The types of values that can be merged are `Object` and `Array`. To set another type of value use `Onyx.set()`. Merge
+ * behavior uses lodash/merge under the hood for `Object` and simple concatenation for `Array`. However, it's important
+ * to note that if you have an array value property on an `Object` that the default behavior of lodash/merge is not to
+ * concatenate. See here: https://github.com/lodash/lodash/issues/2872
+ *
+ * Calls to `Onyx.merge()` are batched so that any calls performed in a single tick will stack in a queue and get
+ * applied in the order they were called. Note: `Onyx.set()` calls do not work this way so use caution when mixing
+ * `Onyx.merge()` and `Onyx.set()`.
+ *
+ * @example
+ * Onyx.merge(ONYXKEYS.EMPLOYEE_LIST, ['Joe']); // -> ['Joe']
+ * Onyx.merge(ONYXKEYS.EMPLOYEE_LIST, ['Jack']); // -> ['Joe', 'Jack']
+ * Onyx.merge(ONYXKEYS.POLICY, {id: 1}); // -> {id: 1}
+ * Onyx.merge(ONYXKEYS.POLICY, {name: 'My Workspace'}); // -> {id: 1, name: 'My Workspace'}
+ *
+ * @param {String} key ONYXKEYS key
+ * @param {(Object|Array)} value Object or Array value to merge
+ * @returns {Promise}
+ */
+declare function merge<K extends KeyList>(key: K, value: Partial<Values[K]>): Promise<void>;
+
+/**
+ * Clear out all the data in the store
+ *
+ * Note that calling Onyx.clear() and then Onyx.set() on a key with a default
+ * key state may store an unexpected value in Storage.
+ *
+ * E.g.
+ * Onyx.clear();
+ * Onyx.set(ONYXKEYS.DEFAULT_KEY, 'default');
+ * Storage.getItem(ONYXKEYS.DEFAULT_KEY)
+ *     .then((storedValue) => console.log(storedValue));
+ * null is logged instead of the expected 'default'
+ *
+ * Onyx.set() might call Storage.setItem() before Onyx.clear() calls
+ * Storage.setItem(). Use Onyx.merge() instead if possible. Onyx.merge() calls
+ * Onyx.get(key) before calling Storage.setItem() via Onyx.set().
+ * Storage.setItem() from Onyx.clear() will have already finished and the merged
+ * value will be saved to storage after the default value.
+ *
+ * @param keysToPreserve is a list of ONYXKEYS that should not be cleared with the rest of the data
+ */
+declare function clear<K extends KeyList>(keysToPreserve?: K[]): Promise<void>;
+
+/**
+ * Merges a collection based on their keys
+ *
+ * @example
+ *
+ * Onyx.mergeCollection(ONYXKEYS.COLLECTION.REPORT, {
+ *     [`${ONYXKEYS.COLLECTION.REPORT}1`]: report1,
+ *     [`${ONYXKEYS.COLLECTION.REPORT}2`]: report2,
+ * });
+ *
+ * @param {String} collectionKey e.g. `ONYXKEYS.COLLECTION.REPORT`
+ * @param {Object} collection Object collection keyed by individual collection member keys and values
+ * @returns {Promise}
+ */
+declare function mergeCollection<K extends KeyList>(collectionKey: K, collection: Record<K, Values[K]>): Promise<void>;
```

Comment:
Even if we type it as:
```ts
declare function mergeCollection<K extends ValueOf<OnyxKeysMap['COLLECTION']>>(collectionKey: K, collection: Record<`${K}${string}`, OnyxValues[`${K}${string}`]>): Promise<void>;
```

Typescript allows to assign any string with [`${}`] syntax:

```ts
mergeCollection(ONYXKEYS.COLLECTION.DOWNLOAD, {
    [`${'test'}`]: 'asdasd', // no error
    [`${'test'}_whatever`]: 'asdasd', // no error
    
    [`test_whatever`]: 'asdasd', // error
    test_whatever: 'asdasd', // error
});
```

I'm not sure if we are able to overcome this issue 🤔

Reply:
Hmm I need to test this more, doing either of these both ways are removing any type checking from `mergeCollection`, not sure why yet.

```ts
declare function mergeCollection<K extends Key>(collectionKey: K, collection: Record<`${K}${string}`, Value[`${K}${string}`]>): Promise<void>;

or...

declare function mergeCollection<K extends ValueOf<KeyValueMap['COLLECTION']>>(collectionKey: K, collection: Record<`${K}${string}`, Value[`${K}${string}`]>): Promise<void>;

    Onyx.mergeCollection(ONYXKEYS.COLLECTION.DOWNLOAD, {
        [`${ONYXKEYS.ACCOUNT}${'attachment1'}`]: {url: 'download_url'},
    });

    Onyx.mergeCollection(ONYXKEYS.IS_LOADING_PAYMENT_METHODS, {
        [`${ONYXKEYS.ACCOUNT}${'attachment1'}`]: 1,
    });
```

I'm putting crazy keys and values in `mergeCollection` and it's allowing everything now. I`ll update both PRs soon so you can test better with the other fixes.

Reply:
The problem here is that TypeScript relaxes the type checking slightly when dealing with template literals because it does not exactly know what will be the final type after concatenation. Atm I don't have an alternative approach 😕

Reply:
Typescript infers such literal strings as `[x: string]: string;` - which causes loose typings here -> [playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBAhjAvDA3gKBp+AuGByOPAGgywCNs8zjTMBtAAwBIU9g8BfBgXUvbQ4BuNFACeABwCmMAIJIYYqSABm8YQHp1WAHoB+IA)

Reply:
I think I have a workaround for this! Here is another [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBA0hJQLxQOQBMD2B3AdgGwwEM0B9FKAH1QCcIwNrgyBuAKFdEigDVC8BXCAGckUAN6soUqJlwFiJAFxQhwagEscAczbSoteoyU5+AWwBGEamwC+7TtABihHAGMQAJQiuGaADwAshhoEHhQEAAewBA4aCKqGtoANFAAKincAHyiEnoA2jBQmlAA1vAYAGZpALrKhZHRsSIABgAkYkEheDbtCZpaNs2S0gD8sOFRMXFQbR3BoYNQYzgQAG5WUMrcUAD0O1AVDObqaOGmYKBFaMNSyivr1qw2bKwhrniEtAf8bsDqGDgoKYrFoIABhDB4PDeP4Avz1SZNWDwFKpTIACh8UJh-xwcBAdRSWOhrlhOGUzjcnm8vnhqIyfEEQgK1UyAEplAAFagYUzqIQQPyrDAnTIvYHUUEQ7Gk3Ho9DYfBEUgoFJiKA3GSK+SkE7KFDRVSq3b7ADyME1siVCkNwH1tuNeyg5s1eVmCrkyrINjw6mi1D4zSghBEPhwqlqqAdKSdLs1Trd7Q91pVPr9VkDkYNwmAjv2Vh51CE8f2ibEyZ13qDIagYYj9pzefC1ELxb0CfdVsrKEGWejJubrZLUDLKAMDCYPdt1dDAPrUcbMfzLYYbekHaT46MPeafcXA4Lq6ebNYQA)

cc @hayata-suenaga @fabioh8010

Reply:
wooow didn't know template literals in object keys result in loose typing...

love your solution with `as const` 🎉  @blazejkustra

Reply:
@fabioh8010 @blazejkustra 

When [I tried in TS Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgNQIYBsCuBTA8gMzgF858oIQ4AiGATzGwFp9sBnGKgbgCg6G4AItmxg0WPPgA8AFQB8cALxxpcbAA8Y2AHYATVnAgAjAFbYAxvAD8g4aIw4CMgNoBrbLQiFpAXXkAuZR5uMwgtdjhcADkATQANAGkAUWiAZUVEbjg4ACVcXGkAfSTogKpyCBh49yoAGky4AGFcABlmxIbpAEkogIR6rKbW9q6oouTSkPR0cxhgUKraAtr6ojqSVH0QsJggvmwIrVo1BdYAWVQwdL3PCJiE5JSePcaIKZm5rQX0gAMAEgQxA4pLhDsd3GcLk4qIM2h1upEqL4iP92FBgFoAOZEb5Pej7EFHL5KIQiQESSS4EDAGAU0Enc5gGrUGHDeFUWTyAA+LzeFg+C12eIORzJ+iUfSyTiicWKKQAdLl8mNot5eqQIBAAqj0RjOHBDKgoAEtJgQIZsFBiDxJW5aAE-ghroRpfdUnKWXCou6WrCRpFlciENrMdjVRkslkDQAvAKGDXTVBaa1W7hEIL4TBaPmhOBgNFaSruGSqDTaPQ86bZz7uWQACnqk0rs3m7gCDVeTf57jqWQAbvZsAECWpRU4fNwAJThuBbVivbBy9AQDG1xvvFu0Jn98QTnhEbjcPPowu0Wsu2XeoaeyKX33w5VMpDRgIwKA4Yi7w-5k9nu4Xj1+rerKjMUj7qpq1D4BqtT6oaAQAIwfjwR4Fgsv4yg8Cp5IUoGILBMZwK+75EJ+QA), it worked

[As I suggested in the App repo PR](https://github.com/blazejkustra/App/pull/2#discussion_r1247297575), we can take union types for collection keys and normal keys separately

> ```
> // by the way, we can get separate types for collection keys and other keys as we have methods on Onyx that just accepts collection keys
> 
> interface CustomTypeOptions {
>   normalKey: unknown; // or string
>   collectionKey: unknown; // or string
>   ...
> 
> type CollectionKey =  CustomTypeOptions['collectionKey'];
> type Key = CustomTypeOptions['normalKey'] | CollectionKey;
> ```

Reply:
Thanks @blazejkustra @hayata-suenaga for your suggestions!

@hayata-suenaga for now I'm following @blazejkustra's suggestion as it's working well with how we have been typing the library. Separating the keys between normal and collection ones would indeed add more type-safety to some Onyx functions, but I think it would require more time and refactorings in all our types.

@blazejkustra wdyt?

Reply:
I didn't investigate yet, but if is easy to refactor the typings of all functions to support `keys` and `collectionKey` as separate types we can totally do this now!

Reply:
I'm okay with keeping things as they are for now. When we decide to convert onyx to TypeScript, that's the right time to split these keys. We may also need to rethink how these keys are initialised so that it is easier to work with internally in react-native-onyx (implementation may have to be changed).

Reply:
I just realised that during [onyx presentation](https://drive.google.com/file/d/1jvVk57tIwu0JFGScVBuMfaF1_J_x1zXX/view) (24:30) Tim said this:

> You can have unintended side effects if you include underscore for keys other than collections. That can mess up Onyx.

Meanwhile we have these keys in E/App:
```ts
    // Contains all the private personal details of the user
    PRIVATE_PERSONAL_DETAILS: 'private_personalDetails',
    // Contains the user preference for the LHN priority mode
    NVP_PRIORITY_MODE: 'nvp_priorityMode',

    // Contains the users's block expiration (if they have one)
    NVP_BLOCKED_FROM_CONCIERGE: 'private_blockedFromConcierge',

    // A unique identifier that each user has that's used to send notifications
    NVP_PRIVATE_PUSH_NOTIFICATION_ID: 'private_pushNotificationID',

    // The NVP with the last payment method used per policy
    NVP_LAST_PAYMENT_METHOD: 'nvp_lastPaymentMethod',
```

We should fix probably fix it, or at least make sure it does work with underscore in the middle of the value @fabioh8010 @hayata-suenaga

Reply:
> @fabioh8010 @blazejkustra
> 
> When [I tried in TS Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgNQIYBsCuBTA8gMzgF858oIQ4AiGATzGwFp9sBnGKgbgCg6G4AItmxg0WPPgA8AFQB8cALxxpcbAA8Y2AHYATVnAgAjAFbYAxvAD8g4aIw4CMgNoBrbLQiFpAXXkAuZR5uMwgtdjhcADkATQANAGkAUWiAZUVEbjg4ACVcXGkAfSTogKpyCBh49yoAGky4AGFcABlmxIbpAEkogIR6rKbW9q6oouTSkPR0cxhgUKraAtr6ojqSVH0QsJggvmwIrVo1BdYAWVQwdL3PCJiE5JSePcaIKZm5rQX0gAMAEgQxA4pLhDsd3GcLk4qIM2h1upEqL4iP92FBgFoAOZEb5Pej7EFHL5KIQiQESSS4EDAGAU0Enc5gGrUGHDeFUWTyAA+LzeFg+C12eIORzJ+iUfSyTiicWKKQAdLl8mNot5eqQIBAAqj0RjOHBDKgoAEtJgQIZsFBiDxJW5aAE-ghroRpfdUnKWXCou6WrCRpFlciENrMdjVRkslkDQAvAKGDXTVBaa1W7hEIL4TBaPmhOBgNFaSruGSqDTaPQ86bZz7uWQACnqk0rs3m7gCDVeTf57jqWQAbvZsAECWpRU4fNwAJThuBbVivbBy9AQDG1xvvFu0Jn98QTnhEbjcPPowu0Wsu2XeoaeyKX33w5VMpDRgIwKA4Yi7w-5k9nu4Xj1+rerKjMUj7qpq1D4BqtT6oaAQAIwfjwR4Fgsv4yg8Cp5IUoGILBMZwK+75EJ+QA), it worked 👍
> 
> [As I suggested in the App repo PR](https://github.com/blazejkustra/App/pull/2#discussion_r1247297575), we can take union types for collection keys and normal keys separately
> 
> > ```
> > // by the way, we can get separate types for collection keys and other keys as we have methods on Onyx that just accepts collection keys
> > 
> > interface CustomTypeOptions {
> >   normalKey: unknown; // or string
> >   collectionKey: unknown; // or string
> >   ...
> > 
> > type CollectionKey =  CustomTypeOptions['collectionKey'];
> > type Key = CustomTypeOptions['normalKey'] | CollectionKey;
> > ```

If we remove these ONYXKEYS with underscore an alternative way would be to extract collection keys like this in `react-native-onyx`.

```ts
type Key = TypeOptions['keys'];
type ExtractCollectionKeys<T> = T extends `${infer P}_${string}` ? T : never;
type CollectionKey = ExtractCollectionKeys<Key>;
```

Reply:
@blazejkustra Good catch! There is indeed a logic [here](https://github.com/Expensify/react-native-onyx/blob/main/lib/Onyx.js#L1075) which is based on `_` presence, which is odd because they could have used `isCollectionKey` internal function. Looks like another potential bug.

Reply:
> We may also need to rethink how these keys are initialised so that it is easier to work with internally in react-native-onyx (implementation may have to be changed).


This is just for initialization `Onyx.init`, right?